crates/epaint/src/shape.rs

```diff
@@ -356,48 +356,56 @@ impl Shape {
     }
 
     /// Move the shape by this many points, in-place.
-    pub fn translate(&mut self, delta: Vec2) {
+    pub fn transform(&mut self, transform: TSTransform) {
         match self {
             Self::Noop => {}
             Self::Vec(shapes) => {
                 for shape in shapes {
-                    shape.translate(delta);
+                    shape.transform(transform);
                 }
             }
             Self::Circle(circle_shape) => {
-                circle_shape.center += delta;
+                circle_shape.center = transform * circle_shape.center;
+                circle_shape.radius *= transform.scaling;
             }
             Self::LineSegment { points, .. } => {
                 for p in points {
-                    *p += delta;
+                    *p = transform * *p;
                 }
             }
             Self::Path(path_shape) => {
                 for p in &mut path_shape.points {
-                    *p += delta;
+                    *p = transform * *p;
                 }
             }
             Self::Rect(rect_shape) => {
-                rect_shape.rect = rect_shape.rect.translate(delta);
+                rect_shape.rect = transform * rect_shape.rect;
             }
             Self::Text(text_shape) => {
-                text_shape.pos += delta;
+                text_shape.pos = transform * text_shape.pos;
+                let mut galley = (*text_shape.galley).clone();
+                let scale_transform = TSTransform::from_scaling(transform.scaling);
+                for row in &mut galley.rows {
+                    row.visuals.mesh.transform(scale_transform);
```

Comment:
This can be optimized slightly

```suggestion
                for row in &mut galley.rows {
                    for v in &mut row.visuals.mesh.vertices {
                        v.pos = Pos2::new(transform.scaling * v.pos.x, transform.scaling * v.pos.y);
                    }
```